src/type_parsers/datetime_parser.jl

```
@@ -49,108 +59,155 @@ function _unsafe_datetime(y=0, m=1, d=1, h=0, mi=0, s=0, ms=0)
-    if (pos >= len || (b != UInt8('T') && b != UInt8(' ')))
-        if !(b == delim || b == cq)
-            code |= Parsers.EOF
-            pos += 1
-        end
-        return _unsafe_datetime(year, month, day), code | Parsers.OK, pos
+    if (pos > len) || (b != UInt8('T') && b != UInt8(' '))
+        return _clamped_datetime(year, month, day), code | Parsers.OK, pos
     end
     # ensure there is enough room for at least HH:MM:DD
-    len - pos < 8 && (return _unsafe_datetime(0), code | Parsers.INVALID, len)
+    len - pos + 1 < 8 && (return _unsafe_datetime(0), code | Parsers.INVALID, pos)
     b = buf[pos += 1]
 
     hour = 0
     for _ in 1:2
-        b -= 0x30
-        b > 0x09 && (return _unsafe_datetime(year, month, day), code | Parsers.INVALID, pos)
-        hour = Int(b) + 10 * hour
+        b0 = b - 0x30
+        b0 > 0x09 && (return _unsafe_datetime(year, month, day), code | Parsers.INVALID, pos)
+        hour = Int(b0) + 10 * hour
         b = buf[pos += 1]
     end
-    hour > 24 && (return _unsafe_datetime(year, month, day), code | Parsers.INVALID, pos)
+    hour >= 24 && (return _unsafe_datetime(year, month, day), code | Parsers.INVALID, pos)
     b != UInt8(':') && (return _unsafe_datetime(year, month, day, hour), code | Parsers.INVALID, pos)
     b = buf[pos += 1]
 
     minute = 0
     for _ in 1:2
-        b -= 0x30
-        b > 0x09 && (return _unsafe_datetime(year, month, day, hour), code | Parsers.INVALID, pos)
-        minute = Int(b) + 10 * minute
+        b0 = b - 0x30
+        b0 > 0x09 && (return _unsafe_datetime(year, month, day, hour), code | Parsers.INVALID, pos)
+        minute = Int(b0) + 10 * minute
         b = buf[pos += 1]
     end
-    minute > 60 && (return _unsafe_datetime(year, month, day, hour), code | Parsers.INVALID, pos)
+    minute >= 60 && (return _unsafe_datetime(year, month, day, hour), code | Parsers.INVALID, pos)
```

looks like a bug fix? shall we add a test for it?

Or maybe we were relying on the `Dates.validargs` call below to catch this? either way, worth checking we have a test?

---

Yes, this was guarded against by `validargs` (so not a bugfix). I've added a test for these in the "invalid" test set.

---

Before I added a bunch of quick checks like this so we'd fail early, but I wasn't sure which values were invalid, I thought there were exceptions like leap seconds being validly represented with a 60 etc, so I left it a bit more open than the actual `validargs` did. Now I removed `validargs` altogether and just used the validation it did in these quick checks.